scripts/nn/util.dml

```diff
@@ -216,3 +216,110 @@ threshold = function(matrix[double] X, double thresh)
   out = X > thresh
 }
 
+top_k_row = function(matrix[double] X, integer r, integer k)
+    return (matrix[double] values, matrix[double] indices) {
+  /*
+   * Computes the top k values (i.e. probabilities) and associated
+   * indices (i.e. classes) in the rth row of the input matrix X
+   *
+   * Inputs:
+   * - X: Inputs, of shape (N D).
+   * - r: Input row number of X to look for
+   * - k: Input number of top elements to look for
+   *
+   * Outputs:
+   * - values: The top k values at the rth row, of shape
+   *    (1, k)
+   * - indices: The class indices, of shape (1, k)
+   */
+
+  #TODO: do r & k need to be checked in the valid range
+  row = X[r, ]
+  row_t = t(row)
+  indices = order(target=row_t, by=1, decreasing=TRUE, index.return=TRUE)
+  indices = t(indices)
+  indices = indices[1, 1:k]
+
+  values = matrix(0, rows=1, cols=k)
+  for (i in 1:k) {
+    values[1, i] = row[1, as.scalar(indices[1, i])]
+  }
+}
+
+top_k = function(matrix[double] X, integer k)
+     return (matrix[double] values, matrix[double] indices) {
+  /*
+   * Computes the top k values (i.e. probabilities) and associated
+   * indices (i.e. classes) for the input matrix X.
+   *
+   * Inputs:
+   * - X: Inputs, of shape (N D).
+   * - k: Input number of top elements to look for
+   *
+   * Outputs:
+   * - values: The top k values along a certain dimension, of shape
+   *    (N, k)
+   * - indices: The indices of classes, of shape (N, K)
+   */
+  N = nrow(X)
+  D = ncol(X)
+  values = matrix(0, rows=N, cols=k)
+  indices = matrix(0, rows=N, cols=k)
+
+  parfor (r in 1:N) {
+    [value, index] = top_k_row(X, r, k)
+    values[r, ] = value
+    indices[r, ] = index
+  }
+}
+
+top_k2d = function(matrix[double] X, int k, int C, int Hin, int Win)
+     return (matrix[double] values, matrix[double] indices) {
+  /*
+   * Computes the top k values (i.e. probabilities) and associated
+   * indices (i.e. classes) for the input matrix X.
+   *
+   * Inputs:
+   * - X: Inputs, of shape (N, C*Hin*Win).
+   * - k: Input number of top elements to look for.
+   * - C: Number of input channels (dimensionality of input depth).
+   * Hin: Input height.
+   *  - Win: Input width.
+   *
+   * Outputs:
+   * - values: The top k values along a certain dimension, of shape
+   *    (N, k*Hin*Win)
+   * - indices: The indices of classes, of shape (N, k*Hin*Win)
+   */
+  N = nrow(X)
+  values = matrix(0, rows=N, cols=k*Hin*Win)
+  indices = matrix(0, rows=N, cols=k*Hin*Win)
+
+  # Reshape the input matrix
```

Comment:
Thanks for adding this.  It turns out that @prithvirajsen just added dedicated function in [PR 552](https://github.com/apache/systemml/pull/552) for transposing from `(N, C*H*W)` to `(C, N*H*W)` using a vectorized implementation, which is great.  It should have the same behavior as the looped implementation, but should be much faster inside the engine.  Once that PR is complete, we could update this to use `X_reshape = t(util::transpose_NCHW_to_CNHW(X, C))`, which would return a shape (`N*Hin*Win, C)`.  Then we could transpose the `values_tmp` and `indices_tmp` back to the desired shape with `values = util::transpose_NCHW_to_CNHW(t(values_tmp), N)`.